fhirpipe/extract/mapping.py

```diff
@@ -0,0 +1,349 @@
+    all_cleaning_scripts = defaultdict(list)
+    # cleaning_scripts has the form
+    # {"script1": (["col1", "col3", ...], [static3]),
+    #  "script4": ([col2], [static1, static3, ...]),
+    #  ...}
+    all_merging_scripts = defaultdict(list)
+    if isinstance(tree, dict):
+
+        # If we are not in a leaf, we recurse
+        if "attributes" in tree.keys() and tree["attributes"]:
+            return find_cols_joins_and_scripts(tree["attributes"])
+
+        cols, joins, cleaning, merging = find_cjs_in_leaf(tree)
+
+        all_cols = all_cols.union(cols)
+        all_joins = all_joins.union(joins)
+        dict_concat(all_cleaning_scripts, cleaning)
+        dict_concat(all_merging_scripts, merging)
+
+    # If the current object is a list, we can repeat the same steps as above for each item
+    elif isinstance(tree, list) and len(tree) > 0:
+        for t in tree:
+            (
+                c_children,
+                j_children,
+                cs_children,
+                ms_children,
+            ) = find_cols_joins_and_scripts(t)
+
+            all_cols = all_cols.union(c_children)
+            all_joins = all_joins.union(j_children)
+            dict_concat(all_cleaning_scripts, cs_children)
+            dict_concat(all_merging_scripts, ms_children)
+
+    return all_cols, all_joins, all_cleaning_scripts, all_merging_scripts
+
+
+def find_cjs_in_leaf(leaf):
+    columns = set()
+    joins = set()
+    cleaning_scripts = defaultdict(list)
+    merging_scripts = defaultdict(list)
+
+    # Check if we need to build the object to put in merging_scripts
+    cols_to_merge = ([], []) if leaf["mergingScript"] is not None else None
+
+    for col in leaf["inputColumns"]:
+
+        # If table and column are defined, we have an sql input
+        if col["table"] and col["column"]:
+            column_name = build_col_name(
+                col["table"], col["column"], col["owner"]
+            )
+            columns.add(column_name)
+
+            # If there are joins, add them to the output
+            for join in col["joins"]:
+                source_col = build_col_name(
+                    join["sourceTable"],
+                    join["sourceColumn"],
+                    join["sourceOwner"],
+                )
+                target_col = build_col_name(
+                    join["targetTable"],
+                    join["targetColumn"],
+                    join["targetOwner"],
+                )
+                joins.add((source_col, target_col))
+
+            # If there is a cleaning script
+            if col["script"]:
+                column_name = build_col_name(
+                    col["table"], col["column"], col["owner"]
+                )
+                cleaning_scripts[col["script"]].append(column_name)
+                if cols_to_merge is not None:
+                    cols_to_merge[0].append(
+                        new_col_name(col["script"], column_name)
+                    )
```

Comment:
Instead of reading `cols_to_merge` as tuple, maybe a dict would be simpler: {'columns': [], 'static': []}
In addition, `new_col_name` is not very clear, and do you really want to put together the 2 strings (leave them as a tuple, or as a dict, might be simpler) ?

Reply:
Oooook I see new_col_name is use to create a column in df and vectorize cleaning scripts execution

Reply:
That's it! I don't think it's very elegant but I didn't find any other solution...

Reply:
Maybe just rename new_col_name and put a few comments to de-confuse people :)